### PR TITLE
Migrate from KBuild to CMake

### PR DESCRIPTION
**Pull request by SebastianBoe**
RFC: [Replacing Make/Kbuild with CMake](https://lists.zephyrproject.org/pipermail/zephyr-devel/2017-April/007538.html)
issue: https://github.com/zephyrproject-rtos/zephyr/issues/1488

This PR migrates Zephyr from Kbuild[1] to CMake[0]. This migration is an important step in a larger
effort to make Zephyr easy to use for application developers working on different platforms with different development environment needs.

Simplified, this change retains Kconfig as-is, and replaces all Makefiles with CMakeLists.txt. The DSL-like Make language that KBuild offers is replaced by a set of CMake extentions. These extentions have either provided simple one-to-one translations of KBuild features or introduced new concepts that replace KBuild concepts.

The PR will be merged when all blocker TODOs are complete or the TSC deems it ready.

Review is welcome and may or may not introduce new blocker or non-blocker TODOs.

NB: Github's "Commits" and "Files changed" feature is broken for some reason.
You'll need to use your local git client to review the changes.

# Blocker TODO

- [x] Port DTC overlay
- [x] Verify that all samples are similar enough before and after CMake
- [x] Verify on HW for a small selection of samples
- [x] Update win docs
- [x] Test building with Windows
- [x] Write migration text/announcement for mailinglist
- [x] Remove KBuild makefiles
- [x] Squash history of review fixes
- [x] Port bt-servr
- [x] openocd is not discovered correctly

# Non-blocker TODO
- [ ] Ninja is broken
- [ ] Port or deprecate xcc (Requires RHEL 4 or 5)
- [x] Update mac docs
- [ ] Update build instructions throughout docs using 'make` (~140 uses)
- [x] The CMake dependency is difficult to deal with

# References:

[0] https://cmake.org/overview/
[1] https://www.kernel.org/doc/Documentation/kbuild/makefiles.txt

Earlier PR https://github.com/zephyrproject-rtos/zephyr/pull/1107
Very early Gerrit preview: https://gerrit.zephyrproject.org/r/#/c/12672/

### CODE REVIEW

**Comment by mbolivar** [comment]
Hi! I have a question about device tree overlay support in the CMake port.

In scripts/Makefile.lib, there is support for a pair of variables which may be set by the application (or, I suppose, on the command line), namely DTC_OVERLAY_FILE and DTC_OVERLAY_DIR. These allow for device tree overlays to be put in place on an application- or board-specific basis, respectively.

I'm looking for equivalent support in CMake, since a sample provided by mcuboot and other applications I work on rely on this functionality. However, I don't see support for DT overlays in dts.cmake or anywhere else.

Am I missing something?

**Comment by SebastianBoe** [comment]
> Am I missing something?
You are not missing anything. I missed this feature.
Added to blocker list.

**Review by mbolivar — CHANGES_REQUESTED** [review]
Hi, I'm having an issue flashing led_ws2812 on 96b_carbon:

```
 KBUILD_VERBOSE=1 ninja flash
[3/3] Flashing 96b_carbon
FAILED: zephyr/cmake/flash/CMakeFiles/flash 
cd /home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build && /usr/bin/cmake -E env DFUUTIL_PID=0483:df11 DFUUTIL_ALT=0 DFUUTIL_IMG=/home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build/zephyr/zephyr.bin DFUUTIL_DFUSE_ADDR= O=/home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build/zephyr KERNEL_ELF_NAME=zephyr.elf KERNEL_HEX_NAME=zephyr.hex KERNEL_BIN_NAME=zephyr.bin ARCH=arm BOARD_NAME=96b_carbon /home/mbolivar/.local/share/virtualenvs/zephyr-q52mD-rc/bin/python3 /home/mbolivar/zmp/zephyr/scripts/support/zephyr_flash_debug.py flash /home/mbolivar/zmp/zephyr/scripts/support/dfuutil.sh
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 0483:df11
Run-time device DFU version 011a
Claiming USB DFU Interface...
Setting Alternate Setting #0 ...
Determining device status: state = dfuERROR, status = 10
dfuERROR, clearing status
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 011a
Device returned transfer size 2048
DfuSe interface name: "Internal Flash  "
dfu-util: Only DfuSe file version 1.1a is supported
dfu-util: (for raw binary download, use the --dfuse-address option)
dfu-util -d,0483:df11 -l
dfu-util -d,0483:df11 -s :leave -a 0 -D /home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build/zephyr/zephyr.bin
Traceback (most recent call last):
  File "/home/mbolivar/zmp/zephyr/scripts/support/zephyr_flash_debug.py", line 50, in <module>
    run(sys.argv[2], sys.argv[1], debug)
  File "/home/mbolivar/zmp/zephyr/scripts/support/zephyr_flash_debug.py", line 39, in run
    runner.run(command)
  File "/home/mbolivar/zmp/zephyr/scripts/support/runner/dfu.py", line 76, in run
    self.check_call(cmd)
  File "/home/mbolivar/zmp/zephyr/scripts/support/runner/core.py", line 257, in check_call
    subprocess.check_call(cmd)
  File "/usr/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['dfu-util', '-d,0483:df11', '-s', ':leave', '-a', '0', '-D', '/home/mbolivar/zmp/zephyr/samples/drivers/led_ws2812/build/zephyr/zephyr.bin']' returned non-zero exit status 74.
ninja: build stopped: subcommand failed.
```

The root cause from the above trace is that DFUUTIL_DFUSE_ADDR is set to the empty string.

I checked boards/arm/96b_carbon/board.cmake, which contains the following:

```
set(FLASH_SCRIPT dfuutil.sh)

set_property(GLOBAL APPEND PROPERTY FLASH_SCRIPT_ENV_VARS
  DFUUTIL_PID=0483:df11
  DFUUTIL_ALT=0
  DFUUTIL_IMG=${PROJECT_BINARY_DIR}/${KERNEL_BIN_NAME}
  DFUUTIL_DFUSE_ADDR=${CONFIG_FLASH_BASE_ADDRESS}
  )
```

It would thus appear that CONFIG_ options are not available to CMake when board.cmake is included. That's strange, though, because kconfig.cmake is included earlier in the boilerplate file.

Any idea what's wrong?

**Comment by mbolivar** [comment]
> You are not missing anything. I missed this feature.
> Added to blocker list.

Awesome, thank you.

**Comment by SebastianBoe** [comment]
@mbolivar : Found the root cause. Will push a fix soon.

**Review by mbolivar — COMMENTED** [review]
Sounds good; thanks for the fast response.

**Comment by SebastianBoe** [comment]
Fix pushed.

**Comment by mbolivar** [comment]
Another question: I'm still building documentation locally with 'make htmldocs' from top level as described in doc/readme.rst. Is it planned to port this to CMake as well? I don't see this in either TODO list.

**Comment by SebastianBoe** [comment]
> Another question: I'm still building documentation locally with 'make htmldocs' from top level as described in doc/readme.rst. Is it planned to port this to CMake as well? I don't see this in either TODO list.

Short answer; no.

The focus is on improving the Windows support and development environment support for application developers. There are no immediate plans to port Zephyr infrastructure that application developers are not expected to be using.

**Comment by mbolivar** [comment]
> Short answer; no.

That makes sense. So we'll just run `make` from the `doc` directory in the future, I'm guessing?

**Comment by nashif** [comment]
> That makes sense. So we'll just run make from the doc directory in the future, I'm guessing?

yep, that is easy enough I guess.

**Review by mbolivar — APPROVED** [review]

**Comment by mbolivar** [comment]
I couldn't find a way to simply dismiss my review, but I'm no longer requesting changes.

**Comment by SebastianBoe** [comment]
@mbolivar : If you were able to do any HW testing then I would be interested to know the results.

**Comment by mbolivar** [comment]
@SebastianBoe I just did a quick sanity check that I could build and flash led_ws2812 on 96b_carbon and that the console output matched what I expected. This was enough to confirm for me that the configurations were being merged correctly, that it could find a SPI device, and that the flashing script glue was OK.

I figure you and @nashif are testing Nordic and Intel parts, so I just wanted to give a quick look at STM32 until the DT overlay support is added. Unfortunately the applications I care most about rely on that feature to build, so I won't be able to do the "real" testing I'd like to until that's at feature parity in CMake.

**Review by nashif — CHANGES_REQUESTED** [review]
blocking merge until we are ready...

**Review by mbolivar — CHANGES_REQUESTED** [review]
I can't build samples/hello_world on qemu_x86. The error when I run ninja is:

```
/home/mbolivar/bin/zephyr-sdk-0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: cannot find libzephyr.a
```

A full log is at https://hastebin.com/oheyexokuy.sql.

I can build for qemu_cortex_m3, however.

**Comment by SebastianBoe** [comment]
NB: Github's "Commits" and "Files changed" feature is broken for some reason.
You'll need to use your local git client to review the changes.

**Comment by jhedberg** [comment]
I took a peek at this and it looks mostly good, however it seems drivers/bluetooth is missing the CMake conversion. Is that known and still in the TODO?

**Comment by nashif** [comment]
> I took a peek at this and it looks mostly good, however it seems drivers/bluetooth is missing the CMake conversion. Is that known and still in the TODO?

I think those have been completely missed! fixing now :)

File: doc/getting_started/installation_linux.rst
Comment (dbkinder) [inline doc/getting_started/installation_linux.rst]: Not all these changes for installed files are cmake related?

Comment (SebastianBoe) [reply doc/getting_started/installation_linux.rst]: Right.

The Ubuntu apt install command has changed slightly.

We use --no-install-recommends to have more control over exactly which packages we depend on.

This was tested using Docker, which uses a minimal Ubuntu install. This new apt-get install 
command is safer than the old one because it is explicitly listing some dependencies that
were assumed before.

**Comment by carlescufi** [comment]
Tested latest branch on Windows/MSYS2 against latest docs. No issues.

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
@mbolivar does hello world build for you now?

**Comment by nashif** [comment]
recheck

**Comment by mbolivar** [comment]
> @mbolivar does hello world build for you now?

Yes.

**Review by mbolivar — APPROVED** [review]

**Comment by mbolivar** [comment]
(no longer requesting changes; just dismissing my previous review)

File: doc/zephyr.doxyfile
Comment (dbkinder) [inline doc/zephyr.doxyfile]: I think there's a missing ``\`` on this line, yes?

**Review by dbkinder — DISMISSED** [review]

**Comment by jhedberg** [comment]
Unless I'm missing something, it seems the default board selection on a per-app basis seems to get broken with this (i.e. I always have to specify BOARD=...), as does the configuration file selection based on the board (e.g. several Bluetooth samples had this). Is this still planned to be fixed?

**Comment by jhedberg** [comment]
Correction, seems the configuration file selection based on the board seems to be working, however if I'm not mistaken BOARD=... now needs to be specified every time when you build?

**Comment by nashif** [comment]
> Correction, seems the configuration file selection based on the board seems to be working, however if I'm not mistaken BOARD=... now needs to be specified every time when you build?

yes, this was always an annoyance when a sample that is supposed to work with many boards has a default set to the authors favourite board. Most samples/tests were set to some qemu board, so you can replicate that by setting it in your environment which will give you the same results.

**Comment by nashif** [comment]
> yes, this was always an annoyance when a sample that is supposed to work with many boards has a default set to the authors favourite board. Most samples/tests were set to some qemu board, so you can replicate that by setting it in your environment which will give you the same results.

btw, a sample that really only works on 1 board can still have a default, there is no issue with that, but most samples/tests we have run on multiple boards and architectures, so setting default was misleading

**Comment by jhedberg** [comment]
I'm getting the following when building almost any Bluetooth sample app for qemu_x86. Our smallest app, the beacon gives:

`/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region 'MMU_LIST' overflowed by 20 bytes
`

whereas e.g. our mesh app gives:

`/opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region 'MMU_LIST' overflowed by 8376 bytes
`

What's going on?

**Comment by mbolivar** [comment]
> btw, a sample that really only works on 1 board can still have a default, there is no issue with that, but most samples/tests we have run on multiple boards and architectures, so setting default was misleading

FWIW (not 100% sure this is relevant), I've upstreamed how we address this in our own applications in the samples/drivers/led_ws2812 and led_lpd8806 samples, and this was propagated to cmake.

By merging in board-specific fragments into a high-level prj.conf (rather than prj.conf XOR prj_BOARD.conf), you can add the maintainer's favorite board on equal footing with other boards that people port things to.

Obviously if you can write a single prj.conf that always works, that's great, but if you can't then I think this is a reasonable compromise.

**Comment by nashif** [comment]
> /opt/zephyr-sdk/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region 'MMU_LIST' overflowed by 8376 bytes
> 
> What's going on?

not sure, I do not get those at all...

**Comment by mbolivar** [comment]
I'm seeing a new warning when building the docs (with `make -C doc htmldocs`):

```
../Kconfig.zephyr:11: warning: The symbol KERNELVERSION references the non-existent environment variable KERNELVERSION and will get the empty string as its value. If you're using Kconfiglib via 'make (i)scriptconfig', it should have set up the environment correctly for you. If you still got this message, that might be an error, and you should email ulfalizer a.t Google's email service.
```

Anyone else?

**Comment by nashif** [comment]
> Anyone else?

that was always there...

**Comment by mbolivar** [comment]
> that was always there...

Nope. Wasn't there when running `make htmldocs` from the top level makefile which is now gone, because it exported KERNELVERSION.

**Comment by mbolivar** [comment]
> I'm getting the following when building almost any Bluetooth sample app for qemu_x86.

I'm seeing similar. E.g. samples/net/echo_server gives:

```
/home/mbolivar/bin/zephyr-sdk-0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: zephyr/zephyr_prebuilt.elf section `.noinit."/home/mbolivar/zmp/zephyr/subsys/net/ip/net_pkt.c".3' will not fit in region `MMU_LIST'
/home/mbolivar/bin/zephyr-sdk-0.9.2/sysroots/x86_64-pokysdk-linux/usr/libexec/i586-zephyr-elf/gcc/i586-zephyr-elf/6.2.0/real-ld: region `MMU_LIST' overflowed by 8136 bytes
```

I tried checking out the latest commit in this branch that still has Kbuild and trying that, which is currently 35a411ebd, and I get a different error:

```
In file included from /home/mbolivar/zmp/zephyr/include/arch/x86/linker.ld:43:0,
                 from /home/mbolivar/zmp/zephyr/arch/x86/soc/ia32/linker.ld:46:
/home/mbolivar/zmp/zephyr/include/linker/linker-defs.h:108:2: error: #error "Expected NUM_KERNEL_OBJECT_FILES to be defined"
 #error "Expected NUM_KERNEL_OBJECT_FILES to be defined"
  ^~~~~
/home/mbolivar/zmp/zephyr/Makefile:868: recipe for target 'linker.cmd' failed
```

It's building fine on mainline.

@jhedberg, while we're here, I'm trying to update the documentation that references Make, and I ran into these lines in qemu_setup.rst:

```
   $ cd samples/net/echo_server
   $ make server
```

and

```
   $ cd samples/net/echo_client
   $ make client
```

Are the "server" and "client" targets still needed (and should they still work as targets to `ninja`) in the post-CMake world? If so I can update that once we sort out what's wrong with the build; if not, any recommendations on what to replace them with?

**Comment by nashif** [comment]
> Nope. Wasn't there when running make htmldocs from the top level makefile which is now gone, because it exported KERNELVERSION.

yeah, the version handling and make it available to the docs is missing. Looking into that now.
I will also add a top-level Makefile to keep existing 'make htmldocs' working

**Comment by nashif** [comment]
> I'm seeing similar. E.g. samples/net/echo_server gives

ok, this could be related to recent changes in master that were not correctly rebased I guess. @SebastianBoe  can you please take a look?

**Comment by nashif** [comment]
@SebastianBoe , the generated version.h does not match between master and the cmake PR:

in master:
```
#ifndef _KERNEL_VERSION_H_
#define _KERNEL_VERSION_H_

#define ZEPHYR_VERSION_CODE 67939
#define ZEPHYR_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))

#define KERNELVERSION \
0x01096300
#define KERNEL_VERSION_NUMBER     0x010963
#define KERNEL_VERSION_MAJOR      1
#define KERNEL_VERSION_MINOR      9
#define KERNEL_PATCHLEVEL         99
#define KERNEL_VERSION_STRING     "1.9.99"

#endif /* _KERNEL_VERSION_H_ */
```

in cmake PR:

```
#ifndef _KERNEL_VERSION_H_
#define _KERNEL_VERSION_H_

#define ZEPHYR_VERSION_CODE 65793
#define ZEPHYR_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))

#define KERNELVERSION          65793
#define KERNEL_VERSION_NUMBER  16843008
#define KERNEL_VERSION_MAJOR   1
#define KERNEL_VERSION_MINOR   9
#define KERNEL_PATCHLEVEL      99
#define KERNEL_VERSION_STRING  "1.9.99"

#endif /* _KERNEL_VERSION_H_ */
```

**Comment by nashif** [comment]
ok, version handling fixed I hope: #4778

**Comment by mbolivar** [comment]
I'm also getting unexpected output when I `ninja run` hello_world for qemu_x86:

```
$ ninja run
[91/92] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: qemu32,+nx,+pae
qemu-system-i386: warning: Unknown firmware file in legacy mode: genroms/multiboot.bin
```

When building for qemu_cortex_m3, I see the expected output (Hello World! arm)

**Comment by mbolivar** [comment]
> ok, version handling fixed I hope: #4778

It's working for me here; thanks. Left a couple comments on that PR.

**Comment by mbolivar** [comment]
@SebastianBoe: flash support for nrf52840_pca10056 appears to have regressed, but I'd like to confirm.

A Makefile.board was previously available for that board, and I tested it with the existing flash scripts fairly recently. But there's no board.cmake in this branch to keep that support.

Since this is a Nordic board, I want to ask if this was intentional.

**Review by jhedberg — CHANGES_REQUESTED** [review]
The original samples/bluetooth/mesh_demo had the following in its Makefile:

ifneq ($(NODE_ADDR),)
CFLAGS += -DNODE_ADDR=$(NODE_ADDR)
endif

This is an essential part of the app since it doesn't rely on external provisioning, i.e. every image needs a unique address. Basically you would always build it using "make NODE_ADDR=0x0001" or similar. Unless I'm missing something, I don't see this taken into account in the cmake migration (I'd propose a patch but my cmake knowledge is still way too limited).

File: samples/net/rpl-node/Makefile
Comment (jukkar) [inline samples/net/rpl-node/Makefile]: There is no CMakeLists.txt file for rpl-node sample.

**Comment by jhedberg** [comment]
The MMU_LIST overflow doesn't seem to happen with the default make generator, but does happen when I use -GNinja.

Another observation: I didn't at first realize that BOARD must be part of the environment during both the cmake and the make step. I was initially only giving it to cmake, and then running make without the environment variable would cause the build to fail with slightly confusing errors.

**Comment by jukkar** [comment]
> Are the "server" and "client" targets still needed (and should they still work as targets to ninja) in the post-CMake world? If so I can update that once it's sorted out what's wrong with the build; if not, any recommendations on what to replace them with?

I am hoping I would be able to run two qemu's against each other after this migration. But it seems that samples/net/common/Makefile.ipstack is been removed and there is not substitute using cmake. This is major issue for testing. Also this prevents simulated 802.15.4 environment using two qemu's from working.

**Comment by jhedberg** [comment]
Question: how do I run qemu with the make generator? I can successfully build an image with that generator using simply "make", but "make run" doesn't seem to work. The new docs say that "ninja run" should work, but as I've previously mentioned ninja doesn't work for me on any Bluetooth sample because of the MMU_LIST overflow issue.

**Comment by jukkar** [comment]
I am using Fedora 26 and I have cmake 3.9.1 installed. Could the cmake be part of SDK and zephyr to use that directly so that user would not need to downgrade/fiddle with cmake sources?

**Comment by jukkar** [comment]
> But it seems that samples/net/common/Makefile.ipstack is been removed and there is not substitute using cmake. This is major issue for testing. Also this prevents simulated 802.15.4 environment using two qemu's from working.

After looking to cmake/qemu directory, the "make server" and "make client" functionality is still there. I was not able to test that yet because of too new cmake version I have.

**Comment by SebastianBoe** [comment]
> Question: how do I run qemu with the make generator? I can successfully build an image with that generator using simply "make", but "make run" doesn't seem to work. The new docs say that "ninja run" should work, but as I've previously mentioned ninja doesn't work for me on any Bluetooth sample because of the MMU_LIST overflow issue. -- @jhedberg 

You are doing it correctly. "make run" should work. If you have found a corner where it doesn't please provide steps to reproduce. Run "make usage" to see all available command line targets.

File: samples/net/rpl-node/Makefile
Comment (SebastianBoe) [reply samples/net/rpl-node/Makefile]: Are you familiar with this sample, or did you just happen to notice it?

Any assistance here would be appreciated.
https://github.com/zephyrproject-rtos/zephyr/issues/4791

Comment (SebastianBoe) [reply samples/net/rpl-node/Makefile]: Ported.

Now it is broken in both KBuild and CMake, whereas before it was absent in CMake and broken in KBuild.

https://github.com/zephyrproject-rtos/zephyr/pull/4792

File: doc/getting_started/installation_linux.rst
Comment (jukkar) [inline doc/getting_started/installation_linux.rst]: Instead of putting the PATH to .bashrc, better would be to direct to `$HOME/.zephyrrc` as then the cmake would not conflict with system installed cmake if that is some newer version etc. The .zephyrrc is sourced by zephyr-env.sh.

**Comment by erwango** [comment]
Seems that we lost the use of SDK openocd.
When running "make flash", it uses my system openocd (older than the SDK one). Then for boards having flash dependencies to SDK version, flash doesn't work.

**Comment by jukkar** [comment]
> After looking to cmake/qemu directory, the "make server" and "make client" functionality is still there. I was not able to test that yet because of too new cmake version I have.

Verifying that "make server" and "make client" works for net echo-server and echo-client samples. I was not able to use ninja as I get the MMU_LIST overflow issue.

**Comment by erwango** [comment]
Another point:
This is surely a commonly admitted cmake way of working and I just need to get used to it, but it feels a bit painful to type 4 commands instead of one simple "make BOARD=<board>".
Is that foreseen to have an alias to at least avoid the "mkdir build/cd build"?

**Comment by ydamigos** [comment]
Running cmake fails to create the needed files (make or ninja) to build sample subsys/usb/cdc_acm for 96b_carbon because the options  USB_DEVICE_VID and USB_DEVICE_PID don't have a default value. Running menuconfig is not possible. Is it possible to build the sample without changing the prj.conf file?

**Review by erwango — CHANGES_REQUESTED** [review]
Seems that we lost the use of SDK openocd.
When running "make flash", it uses my system openocd (older than the SDK one). Then for boards having flash dependencies to SDK version, flash doesn't work.

Last, booting banner disappeared. Not a blocker, but it is a nice feature to see flashed binary is the epxected one (last build):
"***** BOOTING ZEPHYR OS v1.9.99 - BUILD: Nov  8 2017 11:07:46 *****"

**Comment by SebastianBoe** [comment]
>Another point:
This is surely a commonly admitted cmake way of working and I just need to get used to it, but it feels a bit painful to type 4 commands instead of one simple "make BOARD=".
Is that foreseen to have an alias to at least avoid the "mkdir build/cd build"?

It is painful indeed. I use this alias personally for clean builds:

alias cb='rm -rf b && cmake -Bb -H. && cmake --build b'

**Comment by SebastianBoe** [comment]
> Running cmake fails to create the needed files (make or ninja) to build sample subsys/usb/cdc_acm for 96b_carbon because the options USB_DEVICE_VID and USB_DEVICE_PID don't have a default value. Running menuconfig is not possible. Is it possible to build the sample without changing the prj.conf file?

Fixed in https://github.com/zephyrproject-rtos/zephyr/pull/4795

**Comment by erwango** [comment]
>> It is painful indeed. I use this alias personally for clean builds:

> alias cb='rm -rf b && cmake -Bb -H. && cmake --build b'

Thanks for the hint! Wouldn't it be valuable to provide a full set of official zephyr aliases to avoid issues like build issues do to lack of proper build clean and alike?

File: ext/hal/st/stm32cube/CMakeLists.txt
Comment (erwango) [inline ext/hal/st/stm32cube/CMakeLists.txt]: nitpick: alignment

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by SebastianBoe** [comment]
> This is an essential part of the app since it doesn't rely on external provisioning, i.e. every image needs a unique address. Basically you would always build it using "make NODE_ADDR=0x0001" or similar. Unless I'm missing something, I don't see this taken into account in the cmake migration (I'd propose a patch but my cmake knowledge is still way too limited).
@jhedberg 

Fixed.
https://github.com/zephyrproject-rtos/zephyr/pull/4797

**Comment by SebastianBoe** [comment]
> Thanks for the hint! Wouldn't it be valuable to provide a full set of official zephyr aliases to avoid issues like build issues do to lack of proper build clean and alike?

Perhaps we could do so in the future. For now users will have to figure this out on their own.

File: ext/hal/st/stm32cube/CMakeLists.txt
Comment (SebastianBoe) [reply ext/hal/st/stm32cube/CMakeLists.txt]: Fixed all comments in this file:
https://github.com/zephyrproject-rtos/zephyr/pull/4798

**Comment by carlescufi** [comment]
@erwango aliases would be great. Would you give it a go in a PR to this branch?

**Review by galak — CHANGES_REQUESTED** [review]
In the 'Introduce cmake-based rewrite of KBuild' would be nice to have simple usage details and point to the full docs in the commit message.

**Review by jhedberg — APPROVED** [review]

**Review by dbkinder — APPROVED** [review]

**Comment by erwango** [comment]
@carlescufi: I'll have a try

**Comment by mbolivar** [comment]
>> After looking to cmake/qemu directory, the "make server" and "make client" functionality is still there. I was not able to test that yet because of too new cmake version I have.
> Verifying that "make server" and "make client" works for net echo-server and echo-client samples. I was not able to use ninja as I get the MMU_LIST overflow issue.

@jukkar -- OK, so "make server" and "make client" works for Makefile generated systems but not Ninja ones. I've heard on IRC that we're going to start with Make and switch to Ninja when that's fixed.

We should figure out what to document at first, then. I'm working on porting all the board docs that reference "make flash" etc to the new system (https://github.com/zephyrproject-rtos/zephyr/pull/4805). I can start this with make and switch to Ninja in a follow on after it's working, or I can stick with Ninja. This also affects the other PRs that are updating other documents.

I'd like to get a verdict on what to do -- @nashif, @SebastianBoe, @carlescufi could you please let me know to start with Make or Ninja?

**Comment by pfalcon** [comment]
> could you please let me know to start with Make or Ninja?

My 2 cents: people used make before, of course people will use make after. So, it's kinda strange that ninja is even being mentioned when the current goal is to get CMake into the 1.10 release. Please make make work, anything else should wait, thanks ;-).

**Comment by SebastianBoe** [comment]
@mbolivar start with Make.

**Comment by nashif** [comment]
> I'd like to get a verdict on what to do -- @nashif, @SebastianBoe, @carlescufi could you please let me know to start with Make or Ninja?

make

**Comment by carlescufi** [comment]
@mbolivar make, but do not remove the ninja code from the Sphinx script or the doc, simply default to make for now.

**Comment by mbolivar** [comment]
ACK

**Comment by carlescufi** [comment]
@galak yep, I agree with that.

**Comment by mbolivar** [comment]
> @SebastianBoe: flash support for nrf52840_pca10056 appears to have regressed, but I'd like to confirm.

Any comment on this one?

**Comment by SebastianBoe** [comment]
> Any comment on this one? --  @mbolivar 

Fixed. https://github.com/zephyrproject-rtos/zephyr/pull/4811

**Comment by SebastianBoe** [comment]
> In the 'Introduce cmake-based rewrite of KBuild' would be nice to have simple usage details and point to the full docs in the commit message. -- @galak 

Is this msg ok?
https://github.com/SebastianBoe/zephyr/commit/f5b0face7a9cf4d0ee1b3b717ba9992ae37ec9c3

**Comment by SebastianBoe** [comment]
@erwango : Please click the approve button if you approve of the requested changes.

**Comment by carlescufi** [comment]
Commit message is fine, but please use the conventions in the getting started files. Single line mkdir && cd and -DBOARD. Also link to getting started on docs.zephyrproject.org

**Comment by SebastianBoe** [comment]
> Commit message is fine, but please use the conventions in the getting started files. Single line mkdir && cd and -DBOARD. Also link to getting started on docs.zephyrproject.org

Done (on it's way).

**Comment by galak** [comment]
Can we hide zephyr_library() inside something like zephyr_library_sources()?

Looking at an example in boards/arm/96b_carbon/CMakeLists.txt

**Comment by galak** [comment]
Why do we have some boards with zephyr_library() and some just have zephyr_sources()?

boards/arm/cc3220sf_launchxl/CMakeLists.txt vs boards/arm/96b_carbon/CMakeLists.txt

**Comment by mbolivar** [comment]
> Any comment on this one? -- @mbolivar
> Fixed. #4811

Thanks @SebastianBoe!

**Review by carlescufi — APPROVED** [review]

**Comment by nashif** [comment]
> Why do we have some boards with zephyr_library() and some just have zephyr_sources()?
> 
> boards/arm/cc3220sf_launchxl/CMakeLists.txt vs boards/arm/96b_carbon/CMakeLists.txt

Fixed in #4820

File: CONTRIBUTING.rst
Comment (andrewboie) [inline CONTRIBUTING.rst]: i really hope we didn't purge the USE_CCACHE variable...
It might be better to direct users to that instead of having to pass "--ccache" every time

**Review by tyrantlizard77 — APPROVED** [review]

File: CONTRIBUTING.rst
Comment (SebastianBoe) [reply CONTRIBUTING.rst]: We did. I think I know what kind of behaviour you want. I'll push a fix.

Comment (SebastianBoe) [reply CONTRIBUTING.rst]: Fixed: https://github.com/zephyrproject-rtos/zephyr/pull/4837

**Pull request merged**
